### PR TITLE
Adds generics annotations

### LINKED ISSUE #72: Add support for generics via static analysis
As was done in doctrine/collections, we should add support for generics using the `@psalm-template` annotations for generic classes like Repositories and Registries. We should also ensure that generics in PHPStan are supported as well.

See https://github.com/doctrine/collections/pull/177 for the reference implementation in collections.

### PR DESCRIPTION
**Pull request by weph**
Not sure if that's enough to close #72 but it's a start ;)

### CODE REVIEW

**Comment by Ocramius** [comment]
`findAll` and `findBy` should return `list<T>`, I think
Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

**Comment by weph** [comment]
Yeah, wasn't sure about that. It's a list in doctrine/orm, but doctrine/mongodb-odm uses the document id as array key. Not sure about the other odm implementations. I agree that it should be a list, though. Overriding the annotations in odm doesn't seem right but that's just my humble opinion.

**Comment by greg0ire** [comment]
Nice PR! You gave me the motivation to setup Psalm on this lib! See #105

**Comment by weph** [comment]
Do you have any opinions on this one, @alcaeus ?

**Comment by alcaeus** [comment]
> Do you have any opinions on this one, @alcaeus ?

Sorry for the delay. Since ODM doesn't return a list but rather a struct, this has to be `T[]` for the time being.

**Comment by alcaeus** [comment]
Note that this should go into 1.4.x since that version is supposed to have feature parity with 2.0.

@weph could you please change the target branch and rebase? Thanks!

**Comment by greg0ire** [comment]
Psalm support just landed on 1.4.x BTW :)

**Comment by greg0ire** [comment]
I did the rebase for you.

**Comment by greg0ire** [comment]
Reopening to get a report from Travis

**Comment by greg0ire** [comment]
Thanks @weph !

**Comment by weph** [comment]
Thank you, @greg0ire !

File: lib/Doctrine/Persistence/ObjectManager.php [hunk 1]
```diff
@@ -111,9 +115,9 @@ public function flush();
     /**
      * Gets the repository for a class.
      *
-     * @param string $className
```
Comment (andrew-demb) [inline lib/Doctrine/Persistence/ObjectManager.php:114]: I think this line should be preserved to allow IDE know about required argument type.

Comment (alcaeus) [reply lib/Doctrine/Persistence/ObjectManager.php:114]: I agree. Do you mind creating a pull request?

Comment (andrew-demb) [reply lib/Doctrine/Persistence/ObjectManager.php:114]: Here it is #111

**Pull request merged**